### PR TITLE
TLS Definition 2.0

### PR DESCRIPTION
**Pull request by GabrielKlyatis**

### CODE REVIEW

File: netlab/L5/tls_protocol_layer.hpp [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+#include <string>
+#include <ctime>
+
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+
+namespace netlab {
+
+/************************************************************************/
+/*                         TLS Record Layer                             */
```
Comment (nivshani) [inline netlab/L5/tls_protocol_layer.hpp:12]: you are including "tls_definition.hpp" which already includes all the other thing(vector, array, etc) please remove unnecessary include .

File: netlab/L5/tls_protocol_layer.hpp [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+		/* Constructor */
+		TLSRecordLayer()
+			: protocol_version{ 3, 3 },
+			content_type(TLS_CONTENT_TYPE_MAX_VALUE),
+			length(0),
+			fragment{}
+		{ }
+
+		/* Method used for serializing the record layer data into a string type for sending. */
+		std::string serialize_record_layer_data() {
```
Comment (nivshani) [inline netlab/L5/tls_protocol_layer.hpp:32]: you have a lot of implementation in the .hpp file.
i think it is batter to split definition and implantation and move implementation to tls_protocl_layer.cpp.

File: netlab/L5/tls_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+void append_size_to_string(std::string& str, std::size_t value) {
+	for (size_t i = 0; i < sizeof(std::size_t); ++i) {
+		str.push_back(static_cast<char>((value >> (i * 8)) & 0xFF));
+	}
+}
+
+/* Generates a random byte array. */
+template <std::size_t N>
+std::array<uint8_t, N> generate_random_bytes() {
```
Comment (nivshani) [inline netlab/L5/tls_utils.hpp:20]: can we get a reference to array as argument and fill the array ?

File: netlab/L5/tls_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	std::array<uint8_t, N> random_bytes;
+	for (std::size_t i = 0; i < N; i++) {
+		random_bytes[i] = rand() % 256;
```
Comment (nivshani) [inline netlab/L5/tls_utils.hpp]: magic number

File: netlab/L5/tls_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	}
+	return random_bytes;
+}
+
+uint32_t read_uint32_from_iterator(std::string::const_iterator& it) {
+	uint32_t vector_size = 0;
+	for (int i = 0; i <= 3; ++i) {
```
Comment (nivshani) [inline netlab/L5/tls_utils.hpp]: magic number, also < 4 is more clear in my opinion.
maybe even i < sizeof(uint32_t) or maybe use a define

File: netlab/L5/tls_utils.hpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+		vector_size |= static_cast<uint8_t>(*it) << (i * 8);
+		++it;
+	}
+	return vector_size;
+}
+
+uint16_t read_uint16_from_iterator(std::string::const_iterator& it) {
```
Comment (nivshani) [inline netlab/L5/tls_utils.hpp:37]: same as read_uint32_from_iterator.
can it be one method?

File: netlab_testing/TLS_Impl_test.cpp [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+	handshakeProtocol.handshake.body.clientHello.setClientHello();
+
+	std::string serialized_string = handshakeProtocol.serialize_handshake_protocol_data(msg_type);
+
+	handshakeProtocol.deserialize_handshake_protocol_data(serialized_string, msg_type);
+
+	std::string serialized_string_2 = handshakeProtocol.serialize_handshake_protocol_data(msg_type);
+
+	ASSERT_EQ(serialized_string, serialized_string_2);
```
Comment (nivshani) [inline netlab_testing/TLS_Impl_test.cpp:28]: i think this is not perfect.
if you are consistent in the error, like byte order, in both serialize and de-esrlisdfsad you will pass the test but fail in a real communication

File: netlab/L5/tls_utils.hpp (hunk 3 above)
Comment (GabrielKlyatis) [reply netlab/L5/tls_utils.hpp:20]: Sure, the returned array is created inside, and can be created outside and passed as argument instead. Do you want me to do it?

File: netlab/L5/tls_protocol_layer.hpp (hunk 2 above)
Comment (GabrielKlyatis) [reply netlab/L5/tls_protocol_layer.hpp:32]: Good though. I thought that this is the way to go but felt like it was still pretty sparse to do it yet. But, eventually yes (even if no more additions would be made to these classes, though I find it hard to believe)

File: netlab/L5/tls_protocol_layer.hpp (hunk 1 above)
Comment (GabrielKlyatis) [reply netlab/L5/tls_protocol_layer.hpp:12]: Thanks for noticing, done

File: netlab/L5/tls_utils.hpp (hunk 3 above)
Comment (GabrielKlyatis) [reply netlab/L5/tls_utils.hpp:20]: Actually, I think it's better to leave it as is, works better with the way members are set in the constructors.

File: netlab/L5/tls_utils.hpp (hunk 4 above)
Comment (GabrielKlyatis) [reply netlab/L5/tls_utils.hpp]: Fixed

File: netlab/L5/tls_utils.hpp (hunk 5 above)
Comment (GabrielKlyatis) [reply netlab/L5/tls_utils.hpp]: Fixed

File: netlab/L5/tls_utils.hpp (hunk 6 above)
Comment (GabrielKlyatis) [reply netlab/L5/tls_utils.hpp:37]: Can be one with template as was done with random bytes, but I figured it's not very urgent

File: netlab_testing/TLS_Impl_test.cpp (hunk 7 above)
Comment (GabrielKlyatis) [reply netlab_testing/TLS_Impl_test.cpp:28]: I might not understand the question.
I check serialization/deserialization. By the design of the test, if the object is assembled into a string and disassembled back apart into the same object - we passed.  I checked every byte in the result, what do you think may not be covered?

**Pull request merged**